src/lib/validator.js

```diff
@@ -0,0 +1,68 @@
+/*****
+ License
+ --------------
+ Copyright © 2017 Bill & Melinda Gates Foundation
+ The Mojaloop files are made available by the Bill & Melinda Gates Foundation under the Apache License, Version 2.0 (the "License") and you may not use these files except in compliance with the License. You may obtain a copy of the License at
+ http://www.apache.org/licenses/LICENSE-2.0
+ Unless required by applicable law or agreed to in writing, the Mojaloop files are distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.
+ Contributors
+ --------------
+ This is the official list of the Mojaloop project contributors for this file.
+ Names of the original copyright holders (individuals or organizations)
+ should be listed with a '*' in the first column. People who have
+ contributed from an organization can be listed under the organization
+ that actually holds the copyright for their contributions (see the
+ Gates Foundation organization for an example). Those individuals should have
+ their names indented and be marked with a '-'. Email address can be added
+ optionally within square brackets <email>.
+ * Gates Foundation
+ - Name Surname <name.surname@gatesfoundation.com>
+
+ - Georgi Georgiev <georgi.georgiev@modusbox.com>
+--------------
+ ******/
+
+'use strict'
+
+const Config = require('../lib/config')
+
+const BAD_REQUEST_ERROR_CODE = 400
+const BAD_REQUEST_ERROR_DESC = 'Bad Request'
+const DEFAULT_LAG_SECONDS = 300
+
+const fulfilTransfer = (request) => {
+  let validationPassed = true
+  let errorInformation = {
+    errorCode: BAD_REQUEST_ERROR_CODE,
+    errorDescription: BAD_REQUEST_ERROR_DESC,
+    extensionList: {
+      extension: []
+    }
+  }
+
+  const maxLag = (Config.COMPLETED_TIMESTAMP_MAX_LAG_SECONDS || DEFAULT_LAG_SECONDS) * 1000
+  const completedTimestamp = new Date(request.payload.completedTimestamp)
+  const now = new Date()
+  if (completedTimestamp > now) {
+    errorInformation.extensionList.extension.push({
+      key: 'customValidationError',
+      value: 'child "completedTimestamp" fails because future timestamp was provided'
+    })
+    validationPassed = false
+  } else if (completedTimestamp < now - maxLag) {
+    errorInformation.extensionList.extension.push({
+      key: 'customValidationError',
+      value: 'child "completedTimestamp" fails because provided timestamp is way too far in the past'
```

Comment:
is this the correct error message? please reword into acceptable message

Reply:
For the following request (notice payload.extensionList.extension[0].**key1**):
```
{
	"fulfilment": "f5sqb7tBTWPd5Y8BDFdMm9BJR_MNI4isf8p8n4D5pHA",
	"transferState": "COMMITTED",
	"completedTimestamp": "{{completedTimestamp}}",
	"extensionList": {
		"extension": [{
			"key1": "fulfilmentDetail",
			"value": "This is a fulfilment request"
		}]
	}
}
```
the response is:
```
{
    "errorInformation": {
        "errorCode": 400,
        "errorDescription": "Bad Request",
        "extensionList": {
            "extension": [
                {
                    "key": "joiValidationError",
                    "value": "Supplied key fails to match the required format"
                }
            ]
        }
    }
}
```
This is why I decided it is appropriate for expired completedTimestamp request:
```
{
	"fulfilment": "f5sqb7tBTWPd5Y8BDFdMm9BJR_MNI4isf8p8n4D5pHA",
	"transferState": "COMMITTED",
	"completedTimestamp": "2019-01-01T00:00:00.000+00:00",
	"extensionList": {
		"extension": [{
			"key": "fulfilmentDetail",
			"value": "This is a fulfilment request"
		}]
	}
}
```
to return:
```
{
    "errorCode": 400,
    "errorDescription": "Bad Request",
    "extensionList": {
        "extension": [
            {
                "key": "customValidationError",
                "value": "child \"completedTimestamp\" fails because provided timestamp is way too far in the past"
            }
        ]
    }
}
```
If still consider it as correct. If you don't agree, please provide the full response body, that you think is appropriate.

Reply:
My issues isn't the structure its the wording for value. Sorry I should have been more specific.
"child \"completedTimestamp\" fails because provided timestamp is way too far in the past"
should be 
"completedTimestamp fails because provided timestamp exceeded the maximum timeout duration"

The "child \"completedTimestamp\"" structure comes from when we switched the error messages to Joi and Boom. They need to be revisited when we are doing the error handling framework